hoomd/hpmc/PairPotentialStep.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2009-2024 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#include "PairPotentialStep.h"
+
+namespace hoomd
+    {
+namespace hpmc
+    {
+
+PairPotentialStep::PairPotentialStep(std::shared_ptr<SystemDefinition> sysdef)
+    : PairPotential(sysdef), m_params(m_type_param_index.getNumElements())
+    {
+    }
+
+LongReal PairPotentialStep::computeRCutNonAdditive(unsigned int type_i, unsigned int type_j) const
+    {
+    unsigned int param_index = m_type_param_index(type_i, type_j);
+    size_t n = m_params[param_index].m_r_squared.size();
+    if (n > 0)
+        {
+        return slow::sqrt(m_params[param_index].m_r_squared[n - 1]);
+        }
+    else
+        {
+        return 0;
+        }
+    }
+
+LongReal PairPotentialStep::energy(const LongReal r_squared,
```

Comment:
There may be some tricks we can play in this function to reduce average (not asymptotic) runtime.

For example, we could start by checking if the pair distance is larger or equal to the last value in `m_r_squared` before entering the while loop. I suspect this case will come up often, but it would take many iterations of the binary search to finally realize the distance value is outside the distance range.

We may also be able to lower average runtime by checking if the pair distance is less than the first value in `m_r_squared`, or just returning 0 immediately if `N == 0`.

Reply:
The early exit checks do provide a small but measurable benefit in the case where there is only 1 point in the potential. In the general case, the additional checks will slow the code down somewhat.

Reply:
This potential will frequently be used for square wells, so I think it will be worth it.

Reply:
One more potential speed-up is to use an interpolating version of the binary search algorithm. In that case, you determine `m` according to the relation between `r_squared` and `param.m_r_squared[N-1]`: `size_t m = L*(1-r_squared/param.m_r_squared[N-1]) + R*r_squared/param.m_r_squared[N-1]`.
However, that might only help for large 'N' and when the distances are rather uniformly distributed.

Reply:
I don't think the typical use-case of this will be a large number of points spaced linearly in `r**2`.